### PR TITLE
Search form

### PR DESCRIPTION
**Pull request by nikolina-p**
@kunicmarko20

### CODE REVIEW

File: src/Form/PetForm.php [hunk 1]
```diff
@@ -47,6 +48,14 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
                 'class' => Species::class,
                 'choice_label' => 'speciesName',
             ])
+            ->add('gender', ChoiceType::class, [
```
Comment (kunicmarko20) [inline src/Form/PetForm.php]: what does this have to do with search?

File: src/Entity/Pet.php [hunk 2]
```diff
@@ -24,46 +24,61 @@ class Pet
...
      * @ORM\ManyToOne(targetEntity="App\Entity\User", inversedBy="pets")
      * @ORM\JoinColumn(nullable=false)
+     * @Assert\NotBlank(groups={"new", "edit"})
      */
     private $owner;
 
     /**
      * @ORM\ManyToOne(targetEntity="App\Entity\Species", inversedBy="pets")
      * @ORM\JoinColumn(nullable=false)
+     * @Assert\NotBlank(groups={"new", "edit"})
```
Comment (kunicmarko20) [inline src/Entity/Pet.php]: search related?

File: src/Controller/PetController.php [hunk 3]
```diff
@@ -21,6 +22,27 @@ public function __construct(PetService $petService)
         $this->petService = $petService;
     }
 
+    /**
+     * @Route("/", name="index")
+     */
+    public function index(Request $request): Response
+    {
+        $form = $this->createForm(SearchForm::class, $pet = new Pet());
```
Comment (kunicmarko20) [inline src/Controller/PetController.php]: didn't think about doing it this way

File: src/Repository/PetRepository.php [hunk 4]
```diff
@@ -20,4 +20,39 @@ public function __construct(RegistryInterface $registry)
...
+
+    public function searchPets(Pet $pet)
+    {
+        $queryBuilder = $this->getEntityManager()->createQueryBuilder();
+        $queryBuilder->select('p')
+            ->from('App:Pet', 'p');
+        if ($pet->getSpecies() !== null) {
+            $queryBuilder->where($queryBuilder->expr()->eq(
+                'p.species',
+                "'".$pet->getSpecies()->getId()."'"
```
Comment (kunicmarko20) [inline src/Repository/PetRepository.php]: https://www.doctrine-project.org/projects/doctrine-orm/en/2.6/reference/query-builder.html#binding-parameters-to-your-query

File: src/Repository/PetRepository.php [hunk 5]
```diff
@@ -20,4 +20,39 @@ public function __construct(RegistryInterface $registry)
...
+            ));
+        }
+        if ($pet->getAge() !== null) {
+            $queryBuilder->andWhere($queryBuilder->expr()->eq(
+                'p.age',
+                $pet->getAge()
+            ));
+        }
+        $query = $queryBuilder->getQuery();
+        $result = $query->getResult();
```
Comment (kunicmarko20) [inline src/Repository/PetRepository.php]: do you need $query or $result?

File: src/Form/PetForm.php (hunk 1 above)
Comment (nikolina-p) [reply src/Form/PetForm.php]: I wanted to have more criteria for search and in the first commit I added gender and breed.

File: src/Entity/Pet.php (hunk 2 above)
Comment (nikolina-p) [reply src/Entity/Pet.php]: Because my search form is also made based on Pet class, and in the case of search none of form fields is required, I had to define validation rules in more detail so it works for new Pet, edit Pet and search.

File: src/Form/SearchForm.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+                'label' => 'Age',
+                'scale' => 2,
+                'required' => false
+            ]);
+    }
+
+    public function configureOptions(OptionsResolver $resolver): void
+    {
+        $resolver->setDefaults([
+            'data_class' => Pet::class,
```
Comment (kunicmarko20) [inline src/Form/SearchForm.php]: This is wrong, you should be using a separate DTO for this if you already want to use forms

File: src/Form/PetForm.php (hunk 1 above)
Comment (kunicmarko20) [reply src/Form/PetForm.php]: Still doesn't justify why is this part of this PR and not a new one

File: src/Entity/Pet.php (hunk 2 above)
Comment (kunicmarko20) [reply src/Entity/Pet.php]: and that is wrong, as I mentioned in other comment, you should have a separate class just for this

File: src/Repository/PetRepository.php [hunk 7]
```diff
@@ -21,38 +22,35 @@ public function __construct(RegistryInterface $registry)
...
+        $parameters = [];
         $queryBuilder->select('p')
             ->from('App:Pet', 'p');
-        if ($pet->getSpecies() !== null) {
-            $queryBuilder->where($queryBuilder->expr()->eq(
-                'p.species',
-                "'".$pet->getSpecies()->getId()."'"
-            ));
+
+        if ($petDTO->getSpecies() !== null) {
```
Comment (kunicmarko20) [inline src/Repository/PetRepository.php]: maybe you can add `hasSpecies` on `$petDTO`?

File: src/Repository/PetRepository.php [hunk 8]
```diff
@@ -21,38 +22,35 @@ public function __construct(RegistryInterface $registry)
...
+            $parameters['species'] = $petDTO->getSpecies()->getId();
+            $queryBuilder->where('p.species = :species');
```
Comment (kunicmarko20) [inline src/Repository/PetRepository.php]: there should be `setParameter` IIRC, and you can set it immediately, instead of storing values in a new array.

File: src/Repository/PetRepository.php [hunk 9]
```diff
@@ -21,38 +22,35 @@ public function __construct(RegistryInterface $registry)
...
         $queryBuilder->select('p')
             ->from('App:Pet', 'p');
-        if ($pet->getSpecies() !== null) {
-            $queryBuilder->where($queryBuilder->expr()->eq(
-                'p.species',
-                "'".$pet->getSpecies()->getId()."'"
-            ));
+
+        if ($petDTO->getSpecies() !== null) {
+            $parameters['species'] = $petDTO->getSpecies()->getId();
```
Comment (kunicmarko20) [inline src/Repository/PetRepository.php]: what about `getSpeciesId()` method?

File: src/DTO/PetDTO.php [hunk 10]
```diff
@@ -26,6 +26,11 @@ public function getSpecies(): ?Species
         return $this->species;
     }
 
+    public function getSpeciesId(): ?int
+    {
+        return $this->species->getId();
```
Comment (kunicmarko20) [inline src/DTO/PetDTO.php]: what if `species` is null?
